src/t_list.c

```diff
@@ -33,13 +33,115 @@
  * List API
  *----------------------------------------------------------------------------*/
 
+void listTypeConvertListpack(robj *o, int enc) {
+    serverAssert(o->encoding == OBJ_ENCODING_LISTPACK);
+
+    if (enc == OBJ_ENCODING_LISTPACK) {
+        /* Nothing to do... */
+    } else if (enc == OBJ_ENCODING_QUICKLIST) {
+        quicklist *ql = quicklistCreate();
+        quicklistSetOptions(ql, server.list_max_listpack_size, server.list_compress_depth);
+
+        /* Append listpack to quicklist if it's not empty, otherwise release it. */
+        if (lpLength(o->ptr))
+            quicklistAppendListpack(ql, o->ptr);
+        else
+            lpFree(o->ptr);
+        o->ptr = ql;
+        o->encoding = OBJ_ENCODING_QUICKLIST;
+    } else {
+        serverPanic("Unknown hash encoding");
+    }
+}
+
+/* Check the length and size of a number of objects that will be added to list to see
+ * if we need to convert a listpack to a quicklist. Note that we only check string
+ * encoded objects as their string length can be queried in constant time.
+ *
+ * If callback is given the function is called in order for caller to do some work
+ * before the list conversion. */
+void listTypeTryConvertListpack(robj *o, robj **argv, int start, int end,
+                                beforeConvertCB fn, void *data)
+{
+    if (o->encoding != OBJ_ENCODING_LISTPACK) return;
+
+    size_t sz_limit;
+    unsigned long count_limit;
+    size_t sum = 0;
+
+    for (int i = start; i <= end; i++) {
+        if (!sdsEncodedObject(argv[i]))
+            continue;
+        sum += sdslen(argv[i]->ptr);
+    }
+
+    quicklistSizeAndCountLimit(server.list_max_listpack_size,&sz_limit,&count_limit);
+    if (lpBytes(o->ptr)+sum >= sz_limit || lpLength(o->ptr)+end-start+1 >= count_limit ||
+        !lpSafeToAdd(o->ptr, sum))
+    {
+        /* Invoke callback before conversion. */
+        if (fn) fn(data);
+
+        listTypeConvertListpack(o, OBJ_ENCODING_QUICKLIST);
+    }
+}
+
+/* Check the length and size of a quicklist to see if we need to convert it to listpack.
+ *
+ * If callback is given the function is called in order for caller to do some work
+ * before the list conversion. */
+#define QUICKLIST_CONVERT_THRESHOLD 0.5
```

Comment:
@sundb i noticed you removed the define (due to this comment?), but looks like you didn't add `/2`

Reply:
ohh, i see the comment here: https://github.com/redis/redis/pull/11303#discussion_r998052165
i'm slightly concerned, even if the conversion is cheap, i'm afraid of a resonance effect.
i'm not sure the concern about encoding after rdb loading is that important.
let's keep thinking of it.

Reply:
I'm concerned too. What about a pipeline like PUSH, POP, PUSH, POP, PUSH, POP, ... where each command causes a conversion, is the overhead negligible even in this case?

Reply:
@zuiderkwast You are right, there is significant performance degradation when testing as you say, so let me revert this commit for now.

Reply:
i agree with the re-introduction of the convert threshold, but i'm not certain we wanna keep the explicit rdb format.
i think we can implicitly save the old format of one node quicklist, and i think that when we load it back, we can recognize that it's just one node, and if it confirms with the other thresholds, load it as a listpack instead of quicklist.

> because of the threshold, if we save the listpack as a fake quicklist, we will not be able to determine the encoding before it is saved when rdb loading.
This would go against the fact that we should keep a key in the same encoding when reloading.

@sundb can you explain this? maybe i'm overlooking something.

Reply:
This is because quicklist will only be converted to listpack if there is only one node and size is below the threshold.
If a listpack that size exceeds the threshold, we save it as a fake quicklist, and when reload it, we have no way to know that it's a listpack, we can only treat it as a quicklist that has shrunk to only one node and hasn't reached the threshold.

Reply:
can't we check the size of the listpack on load time, and decide how to handle it?
also, i don't understand the concern of `we should keep a key in the same encoding when reloading`.
our rdb loader changes decides on encoding in load time in many places.
e.g. here (if the config on the instance which saved it is different than the one that loads it)
```c
                if (intsetLen(o->ptr) > server.set_max_intset_entries)
                    setTypeConvert(o,OBJ_ENCODING_HT);
```

Reply:
> also, i don't understand the concern of `we should keep a key in the same encoding when reloading`. 

just like these tests.
https://github.com/redis/redis/blob/e9d4ed4e0fda81aefa0cebf2559dc233d472c902/tests/unit/type/list.tcl#L1441-L1443
https://github.com/redis/redis/blob/e9d4ed4e0fda81aefa0cebf2559dc233d472c902/tests/unit/type/set.tcl#L99-L114
They both assume that after `debug reload`, the encoding of the keys will always be the same as before.
I'm also not sure if there is such a principle.
Without it, I should go on the way as you said.

Reply:
i don't understand that concern, and i see these assertions are VERY old, i think we can show flexibility here, either of these solutions is arguably ok:
1. change OBJECT ENCODING to always say "quicklist", in that case what we say that a single listpack is an optimization of a quicklist (i.e. imagine you would have done your work inside quicklist.c an a special 0 footprint quicklist)
2. change the test to accept both "quicklist" and "listpack"
3. change the test in a way that somehow it doesn't hit this problem (tune something so that size is always either big enough or small enough so that the encoding gets consistent)

from a quick look, i think for both set and list tests you mentioned above, the 3rd option is the right one.

Reply:
Choice the third way, please see commit [`7510886` (#11303)](https://github.com/redis/redis/pull/11303/commits/75108861d9755237647bec04a827156changing050d02ba5).
Another minor change to this commit is changing the limit length of listpack, use `list-max-listpack-size instead` of the original `list-max-listpack-size - 1`.
Because since we treat it as a quicklist node, we should allow it to reach the length of list-max-listpack-size.

Reply:
i can't find that commit.

Reply:
Sorry for that.
Yes, I think the conversion to listpack when growing is mostly useless unless we dynamically modify `list-max-listpack-size`.